### PR TITLE
kata-deploy: docs: say which installations kata-deploy detects 

### PR DESCRIPTION
**Pull request by fidencio**
Detection probes for the systemd units each distribution is known by, so a node running one under another unit name, or even another init system, which would wide way too much what kata-deploy currently does and add a lot of complexity for what we consider edge cases and chasing those edge cases is surely not worth it right now.

When looking at k3s alone (which is the issue reported that triggered this commut), it can be installed under a renamed unit, another systemd directory or openrc, and that's a maintainability nightmare for us.

Refs: https://github.com/kata-containers/kata-containers/issues/13825

### CODE REVIEW

File: docs/helm-configuration.md [hunk 1]
```diff
@@ -203,6 +203,29 @@ leave the kubelet's own `/var/lib/kubelet` alone and need nothing.
...
 
+### criServiceName
+
+The install finds the node's CRI runtime by probing the systemd units each flavour
+is known by — `k3s`, `rke2-server` and so on. A cluster that renamed its unit
+answers to none of them, so name it here:
+
+```yaml title="values.yaml"
+k8sDistribution: k3s
+criServiceName: k3s-custom   # (1)!
```
Comment (manuelh-dev) [inline docs/helm-configuration.md]: would criSystemdUnit or criSystemdUnitName be a better name?

File: tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml [hunk 2]
```diff
@@ -292,6 +292,12 @@ monitor:
 # says nothing about the kubelet, so this value still applies there.
 k8sDistribution: "k8s"  # k8s, k3s, rke2, k0s, microk8s
 
+# The systemd unit running this node's CRI runtime, when it is not the one its
```
Comment (manuelh-dev) [inline tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml]: A  bit clearer, at least to me would be something like:
```
# Override the systemd unit that runs the node's CRI runtime. Kata-deploy uses
# this unit for restarts, readiness checks, and runtime service dependencies.
# When empty, the standard unit name is derived from the detected runtime.
#
# Set this when the CRI runtime is managed by a renamed or custom systemd unit.
# This value does not select the runtime's configuration paths; set
# k8sDistribution separately when using K3s, RKE2, k0s, or MicroK8s.
#
# ".service" is added to a bare name. Specify other unit types in full.
```

File: tools/packaging/kata-deploy/binary/src/runtime/manager.rs [hunk 3]
```diff
@@ -491,13 +504,52 @@ mod tests {
     use std::time::Duration;
     use tempfile::tempdir;
 
+    /// With no unit to probe, the flavour is all that is left (#13825).
```
Comment (manuelh-dev) [inline tools/packaging/kata-deploy/binary/src/runtime/manager.rs]: nit, not sure if we need/want to reference that github issue here - would we not want to do this then with every fix we make, consistently?

File: tools/packaging/kata-deploy/binary/src/runtime/manager.rs [hunk 4]
```diff
@@ -155,9 +161,8 @@ fn known_distribution(distribution: &str) -> Option<&'static str> {
 /// The Kubernetes flavour this node runs, for the parts of the install the CRI
 /// runtime cannot answer for - the kubelet's root directory, say.
 ///
-/// The declaration wins because it describes the cluster rather than inferring
-/// it from one node; failing that, a runtime only one flavour ships names it
-/// just as well.
+/// The runtime can stand in for a missing declaration, since one like
```
Comment (manuelh-dev) [inline tools/packaging/kata-deploy/binary/src/runtime/manager.rs]: 'stand in for a missing declaration' took me quite some time to understand what this means. How about something like:
```
/// Use the explicitly configured Kubernetes distribution when recognized.
/// Otherwise, infer it from a distribution-specific runtime name such as
/// `rke2-agent`. Generic runtimes such as `containerd` identify no distribution.
```

File: tools/packaging/kata-deploy/binary/src/runtime/manager.rs [hunk 5]
```diff
@@ -102,11 +108,11 @@ pub async fn get_container_runtime(config: &Config) -> Result<String> {
     Ok(runtime)
 }
 
-fn declared_runtime_override(distribution: Option<&str>) -> Option<&'static str> {
-    match distribution {
-        Some("microk8s") => Some("microk8s"),
-        _ => None,
-    }
+/// The flavour's first role: only the unit to restart tells the roles apart.
```
Comment (manuelh-dev) [inline tools/packaging/kata-deploy/binary/src/runtime/manager.rs]: I am not sure I comprehend this comment.

File: tests/functional/kata-deploy/kata-deploy-distribution.bats [hunk 6]
```diff
@@ -57,3 +57,24 @@ render() {
...
+	for mode in job daemonset; do
+		rendered=$(render --set "deploymentMode=${mode}" \
+			--set k8sDistribution=k3s \
+			--set criServiceName=k3s-custom)
+
+		echo "${rendered}" | grep -A1 'name: CRI_SERVICE_NAME' | grep -q 'value: "k3s-custom"'
+	done
+}
+
+@test "Helm template: the CRI unit is only reported when named" {
```
Comment (manuelh-dev) [inline tests/functional/kata-deploy/kata-deploy-distribution.bats]: Maybe it's just me, maybe this is better: `Helm template: omits CRI_SERVICE_NAME when criServiceName is unset`

File: docs/helm-configuration.md [hunk 7]
```diff
@@ -203,6 +203,29 @@ leave the kubelet's own `/var/lib/kubelet` alone and need nothing.
     by both. Getting either wrong is quiet: volume updates stop arriving, and GPU
     cold plug falls back to CDI annotations.
 
+### criServiceName
```
Comment (manuelh-dev) [inline docs/helm-configuration.md]: maybe it's my fried Friday brain. In my mind, I am trying to simplify the explanation a bit. How about:

### criServiceName
By default, kata-deploy probes the standard systemd units associated with each
supported runtime. It then derives the unit used for runtime restarts, readiness
checks, and service dependencies.
Set `criServiceName` when the CRI runtime is managed by a renamed or custom
systemd unit:
```yaml title="values.yaml"
k8sDistribution: k3s
criServiceName: k3s-custom   # (1)!
```

1. .service is added to bare names, such as k3s-custom. Specify another systemd unit type in full.

criServiceName overrides the derived systemd unit, but does not select the runtime's configuration layout. Set k8sDistribution separately so kata-deploy uses the correct configuration and kubelet paths.

!!! note "Standard installations"

Leave `criServiceName` empty when the runtime uses its standard systemd unit.
Standard unit probing distinguishes server and agent roles.
If no standard unit can be detected, `k8sDistribution` identifies only the
distribution, not the node role. The current fallback selects the
distribution's first role, so a custom agent or worker unit must be named
explicitly.

File: docs/helm-configuration.md [hunk 8]
```diff
@@ -203,6 +203,29 @@ leave the kubelet's own `/var/lib/kubelet` alone and need nothing.
     by both. Getting either wrong is quiet: volume updates stop arriving, and GPU
```
Comment (manuelh-dev) [inline docs/helm-configuration.md:203]: A general comment: I had a hard time understanding code vs comments and commit message / PR title. The code is actually quite clear and LGTM, but I was first struggling to read it right.

An optional commit message which elaborates a bit more on the actual problem, consequences and solution approach, could be:
```
kata-deploy: support custom CRI systemd units

kata-deploy detects K3s, RKE2, and similar runtimes by probing their
standard systemd units. When the runtime uses a custom unit name, none of
these probes succeeds. Kata-deploy then interprets the generic containerd
version as a vanilla containerd installation, writes configuration to the
wrong location, and attempts to restart containerd.service.
Keep probing standard units to identify the node role. If no standard unit
is detected, use k8sDistribution to select the runtime family and its
configuration layout.
Add the criServiceName Helm value for overriding the systemd unit used by
kata-deploy. When set, this unit is used for restarts, status checks, and
snapshotter service dependencies. When unset, kata-deploy continues deriving
the standard unit from the detected runtime.
Standard installations require no configuration changes. Installations
using a custom CRI unit name must set both values:
- k8sDistribution selects the correct runtime configuration layout.
- criServiceName identifies the custom systemd unit.

Fixes: #13825

Assisted-by: Cursor <cursoragent@cursor.com>
Signed-off-by: Fabiano Fidêncio <ffidencio@nvidia.com>
```

**Comment by manuelh-dev** [comment]
@fidencio - the PR generally looks good to me, and I don't want to block progress. However:
- I have added some comments on readability; at least for those less versed with kata-deploy, this may be helpful. No need to take these comments, for others, this may be easier to read than for me - so taking this into account is up to your discretion.
- I have a question though on a design choice, likely this is an intentional choice, but maybe this might also lack clarity in describing the intent in the PR description, commit message, documentation, or code comments: Why is k8sDistribution treated only as a fallback here? When set to a recognized distribution-specific value (e.g., k3s), it already determines which runtime configuration directory the chart mounts. I would therefore expect that explicit declaration to take precedence (for the runtime family); with unit probing used only to identify the server/agent role or reject a conflicting declaration. Doing so, this would also be consistent with the usage of criServiceName: when supplied, the explicitly configured systemd unit takes *precedence* over the unit derived from runtime detection.

**Comment by fidencio** [comment]
@manuelh-dev, your review made me revisit the k3s documentation, and it's way more complicated than what I've proposed here.

With k3s one can configure it according to the following env vars: https://docs.k3s.io/reference/env-variables ... which not only allows one to change systemd units, but also the default systemd directory, and unit types. This would be a nightmare for us to resolve properly, in a way that I'm really thinking about just documenting, for now, that k3s is only supported on its default locations, and I will re-work the whole PR to do so, and add a big visible note about this in our documentation.

File: docs/helm-configuration.md (hunk 1 above)
Comment (fidencio) [reply docs/helm-configuration.md]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

File: tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml (hunk 2 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/helm-chart/kata-deploy/values.yaml]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

File: tools/packaging/kata-deploy/binary/src/runtime/manager.rs (hunk 4 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/binary/src/runtime/manager.rs]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

File: tools/packaging/kata-deploy/binary/src/runtime/manager.rs (hunk 5 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/binary/src/runtime/manager.rs]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

File: tests/functional/kata-deploy/kata-deploy-distribution.bats (hunk 6 above)
Comment (fidencio) [reply tests/functional/kata-deploy/kata-deploy-distribution.bats]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

File: docs/helm-configuration.md (hunk 7 above)
Comment (fidencio) [reply docs/helm-configuration.md]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

File: docs/helm-configuration.md (hunk 8 above)
Comment (fidencio) [reply docs/helm-configuration.md:203]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

File: tools/packaging/kata-deploy/binary/src/runtime/manager.rs (hunk 3 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/binary/src/runtime/manager.rs]: Marking as resolved, as your review made me revisit the k3s docs and then realize that we should actually simplify things out rather than add complexity for edge cases.

Thanks for the suggestion, please, take a look at the new approach (docs only)!

**Review by manuelh-dev — APPROVED** [review]
ack @fidencio - thanks for documenting this.

**Review by stevenhorsman — APPROVED** [review]
